[move-spec-test] README updated (#82)
eigerco/aptos-core: third_party/move/tools/move-spec-test/README.md

---
id: move-spec-test
title: Move Specification Test tool
---

# Move Specification Test tool

## Summary

This tool is used to test the quality of the Move specifications.

## Overview

The Move Specification Test tool uses the Move Mutator tool to generate mutants
of the Move code. Then, it runs the Move Prover tool to check if the mutants
are killed (so Prover will catch an error) by the original specifications.

> If the mutants are not killed, it means that the specification is incorrect or
> not tight enough to catch such cases.

Reviewer: ```suggestion
If the mutants are not killed, it means that the specification has issues and is
incorrect or not tight enough to catch such cases, so it should be improved.
```

Author: Done.

> Move Specification Test tool works on the Move packages and cannot be run on
> single Move files. It is possible to run the tool on the whole package or
> specific modules only.

Reviewer: ```suggestion
Move Specification Test tool can be used on:
- whole Move packages (projects)
- specific modules only

It cannot be used with single Move files.
```

Author: Done.

The tool generates a report in a JSON format. The report contains information
about the number of mutants tested and killed and also the differences between
the original and modified code.

## Example Usage

Please build the whole repository first. In the `aptos-core` directory, run:
```bash
cargo build --release
```

Then build also the `move-cli` tool:
```bash
cargo build -r -p move-cli
```

Check if the tool is working properly by running its tests:
```bash
cargo test -p move-spec-test
```

The Move Specification Test tool demands the Move Prover to be installed and
configured correctly. Please refer to the Move Prover documentation for more
details.

## Usage

Before checking if the tool works, please make sure that the Move Prover is

> installed and configured correctly. Especially, be sure that all the
> dependencies and backend tools are installed and accessible.

Reviewer: ```suggestion
installed and configured correctly. Especially, ensure all the
dependencies and backend tools are installed and accessible.
```

Author: Done.
